src/rdkafka_cgrp.c

```
@@ -2431,6 +2431,264 @@ static void rd_kafka_cgrp_offset_commit_tmr_cb (rd_kafka_timers_t *rkts,
+                     "in join state %s",
+                     rkcg->rkcg_group_id->str, partitions->cnt,
+                     rkcg->rkcg_assignment->cnt,
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
+
+        if (rd_atomic32_get(&rkcg->rkcg_assignment_lost) &&
+            partitions->cnt != rkcg->rkcg_assignment->cnt) {
+                rd_kafka_log(rkcg->rkcg_rk, LOG_WARNING, "UNASSIGN",
+                             "Group \"%s\": current assignment is lost, but "
+                             "not all partitions were unassigned",
+                             rkcg->rkcg_group_id->str);
+        }
+
+        /* If the consumer has raised a fatal error, remove the entire
+         * assignment */
+        if (rd_kafka_fatal_error_code(rkcg->rkcg_rk)) {
+                if (partitions->cnt < rkcg->rkcg_assignment->cnt)
+                        rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER,
+                                     "UNASSIGN", "Group \"%s\": consumer is "
+                                     "in a failed state, treating incremental "
+                                     "unassign as unassign",
+                                     rkcg->rkcg_group_id->str);
+                return rd_kafka_cgrp_assign(rkcg, NULL);
+        }
+
+        /* verify partitions exist in current assignment before making
+         * any changes to it to ensure the operation doesn't partially
+         * succeed. */
+        for (i = 0; i < partitions->cnt; i++) {
+                rd_kafka_topic_partition_t *rktpar = &partitions->elems[i];
+                if (!rd_kafka_topic_partition_list_find(rkcg->rkcg_assignment,
+                                                        rktpar->topic,
+                                                        rktpar->partition)) {
+                        // MH: TODO: unsure if this warrants a LOG_ERR.
+                        rd_kafka_dbg(rkcg->rkcg_rk,
```

Since this is an error the function should return an error object with a descriptive error message rather than the debug.

---

very good. it's all clear now.